### PR TITLE
IGNITE-21750 Throw a special exception when trying to read from/write to a destroyed IndexStorage

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-21750

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/StorageDestroyedException.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        this("Storage is already destroyed");
+    }
+
+    /**
+     * Constructor.
+     *
+     * @param message Error message.
+     */
+    public StorageDestroyedException(String message) {
+        super(Storage.ALREADY_CLOSED_ERR, message);
```
Comment (tkalkirill) [inline modules/storage-api/src/main/java/org/apache/ignite/internal/storage/StorageDestroyedException.java:41]: Maybe we need a new code for the error, otherwise it's a little strange?

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/StorageStates.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import org.apache.ignite.internal.storage.util.StorageState;
+
+/**
+ * Utilities to work with storage states.
+ */
+public class StorageStates {
+    /**
+     * If not already in a terminal state, transitions to the supplied state and returns {@code true}, otherwise just returns {@code false}.
+     */
+    public static boolean transitionToTerminalState(StorageState targetState, AtomicReference<StorageState> stateRef) {
```
Comment (tkalkirill) [inline modules/storage-api/src/main/java/org/apache/ignite/internal/storage/StorageStates.java:30]: Lets move method to `org.apache.ignite.internal.storage.util.StorageUtils`.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java [hunk 3]
```diff
@@ -364,7 +364,10 @@ private GcQueue createGcQueue(
     CompletableFuture<Void> destroyMvPartitionStorage(AbstractPageMemoryMvPartitionStorage mvPartitionStorage) {
         // It is enough for us to close the partition storage and its indexes (do not destroy). Prepare the data region, checkpointer, and
         // compactor to remove the partition, and then simply delete the partition file and its delta files.
-        mvPartitionStorage.close();
+        boolean transitioned = mvPartitionStorage.transitionToDestroyedState();
```
Comment (tkalkirill) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java:367]: Variable is not needed right?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryTableStorage.java [hunk 4]
```diff
@@ -182,9 +182,14 @@ CompletableFuture<Void> clearStorageAndUpdateDataStructures(AbstractPageMemoryMv
     CompletableFuture<Void> destroyMvPartitionStorage(AbstractPageMemoryMvPartitionStorage mvPartitionStorage) {
         VolatilePageMemoryMvPartitionStorage volatilePartitionStorage = (VolatilePageMemoryMvPartitionStorage) mvPartitionStorage;
 
-        volatilePartitionStorage.transitionToDestroyingState();
+        boolean transitioned = volatilePartitionStorage.transitionToDestroyedState();
```
Comment (tkalkirill) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryTableStorage.java:185]: Variable is not needed right?

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/Index.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import org.apache.ignite.internal.storage.util.StorageState;
+import org.apache.ignite.internal.util.IgniteUtils;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.WriteBatch;
+
+/**
+ * Represents an index for all its partitions.
+ */
+abstract class Index<S extends AbstractRocksDbIndexStorage> {
+    protected final ConcurrentMap<Integer, S> storages = new ConcurrentHashMap<>();
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/Index.java:33]: Lets rename to `storageByPartitionId`.

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/StorageDestroyedException.java (hunk 1 above)
Comment (rpuch) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/StorageDestroyedException.java:41]: To be honest, I'm not sure. Does the user need to know the difference between a destroyed and closed storage? From the user's perspective, it's a storage with which no interactions are possible as it's 'not available anymore'. The existing code seems to be fine for both cases.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java (hunk 3 above)
Comment (rpuch) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java:367]: It is introduced just to make the code more readable, it explains what it means. If I inline it, the code would look like `if mvPartitionStorage.transitionToDestroyedState()`, and it's not obvious right away what would `true` or `false` mean in this case. But if there is a variable called `transitioned`, then it's obvious that the transition we tried was successful.

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/StorageDestroyedException.java (hunk 1 above)
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/StorageDestroyedException.java:41]: It’s at your discretion, of course, but for me it looks like: I want to go home, but I can’t because the house is closed or destroyed, no, it’s closed in any case.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java (hunk 3 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java:367]: For me, the variable does not add more clarity, at your discretion, of course, I would remove it, since the code is not complicated and understandable.

**Review by tkalkirill — APPROVED** [review]
